### PR TITLE
WIP WIP WIP : Buzzermodule

### LINKED ISSUE #105: Buzzer Support
I think we need a buzzer module with the following requirements:

* Configurable output (digital or pwm on something supported by the hardware)
* Configurable alarm conditions (beep on particular alarm conditions)
* User-initiated beeping (accessory is|isn't at a particular value)

Supporting naze's nfet thing would be nice.  [pwm buzzers](http://hobbyking.com/hobbyking/store/__14521__HobbyKing_174_Discovery_Buzzer.html) exist and should be quite easy to support.

See also:  https://github.com/TauLabs/TauLabs/issues/1349

### PR DESCRIPTION
**Pull request by henn1001**
fixes #105 

So far this is only for digital buzzer support and is more or less functional now with these target changes [link](https://github.com/henn1001/dRonin/commit/5014ded9d1bd43465c0f62d2af1f8dc91a24c830)
Can and should probably (maybe?) adapted to pwm type output as well. I´m putting it out here so it can be discussed in which direction this pr should go before i put more work into it.

I made it customizable so people can choose which alarms are active. Options are On|Off|WasArmed.

[Review on Reviewable](https://reviewable.io/reviews/d-ronin/dronin/765)

### CODE REVIEW

File: flight/Modules/Buzzer/buzzer.c [hunk 1]
```diff
@@ -0,0 +1,260 @@
+/**
+ ******************************************************************************
+ * @addtogroup TauLabsModules Tau Labs Modules
+ * @{
+ * @addtogroup BuzzerModule Buzzer Module
+ * @{
+ *
+ * @file       buzzer.c
+ * @author     The OpenPilot Team, http://www.openpilot.org Copyright (C) 2010.
```
Comment (mluessi) [inline flight/Modules/Buzzer/buzzer.c]: If this is new file, written from scratch, only put the dRonin copyright.

File: flight/Modules/Buzzer/buzzer.c [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+#define SAMPLE_PERIOD_MS            200
+// Private types
+
+// Private variables
+static bool module_enabled = false;
+static struct pios_thread *buzzerTaskHandle;
+static int8_t buzzerPin = -1; //Pin for buzzer
+static int8_t accChan = -1; //Accessory Channel
+static bool buzzer_settings_updated;
+static bool wasArmed;
```
Comment (mluessi) [inline flight/Modules/Buzzer/buzzer.c]: Try not to mix CamelCase and underscore (`_`) for variables. In general, use lowercase and `_` for everything except UAVO object related things.

File: flight/Modules/Buzzer/buzzer.c [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+// ****************
+// Private functions
+static void buzzerTask(void * parameters);
+static void getTune(uint8_t arming, uint32_t *newTune);
+
+static int32_t BuzzerStart(void)
+{
+	if (module_enabled) {
+		//FlightBuzzerSettingsConnectCallbackCtx(UAVObjCbSetFlag, &buzzer_settings_updated);
```
Comment (mluessi) [inline flight/Modules/Buzzer/buzzer.c]: Why is this commented out? Looks correct and is needed.

File: flight/Modules/Buzzer/buzzer.c [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+		if (buzzOn)
+			PIOS_GPIO_Off(buzzerPin);  //needs a inversion check for npn/pnp transistor
+		else
+			PIOS_GPIO_On(buzzerPin);
+
+	}
+}
+
+
+static void getTune(uint8_t arming, uint32_t *newTune)
```
Comment (mluessi) [inline flight/Modules/Buzzer/buzzer.c]: Why not just return a `uint32_t`? The pointer usage is correct but returning an uint makes the code simpler IMO.

File: flight/Modules/Buzzer/buzzer.c [hunk 5]
```diff
@@ -0,0 +1,260 @@
+/**
+ ******************************************************************************
+ * @addtogroup TauLabsModules Tau Labs Modules
```
Comment (mluessi) [inline flight/Modules/Buzzer/buzzer.c]: dRonin :)

**Comment by mluessi** [comment]
Nice work @henn1001 Have you had a chance to test this?

File: flight/Modules/Buzzer/buzzer.c [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+	lastSysTime = PIOS_Thread_Systime();
+	
+	while (true) {
+		PIOS_Thread_Sleep_Until(&lastSysTime, SAMPLE_PERIOD_MS);
+
+		if (buzzer_settings_updated) {
+			buzzer_settings_updated = false;
+			FlightBuzzerSettingsGet(&buzzerSettings);
+
+			accChan = buzzerSettings.AcessoryChan;
```
Comment (mluessi) [inline flight/Modules/Buzzer/buzzer.c]: No need to use separate variables for these, just use `buzzerSettings.AcessoryChan` etc directly in the code

File: flight/Modules/Buzzer/buzzer.c [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+
+// Private variables
+static bool module_enabled = false;
+static struct pios_thread *buzzerTaskHandle;
+static int8_t buzzerPin = -1; //Pin for buzzer
+static int8_t accChan = -1; //Accessory Channel
+static bool buzzer_settings_updated;
+static bool wasArmed;
+static bool wasDisarmed;
+static 	FlightBuzzerSettingsData buzzerSettings;
```
Comment (mluessi) [inline flight/Modules/Buzzer/buzzer.c]: Many of the variables above should be allocated inside `buzzerTask`, so the module doesn't use heap if it isn't activated. I think only `module_enabled`, `buzzerTaskHandle`, `buzzer_settings_updated` should be allocated above.

**Comment by henn1001** [comment]
Everything seems to work on my bench. So yeah, i tested it. I´ll dig through your comments later and fix the issues.
I mostly put it out here to get some discussion going how we want to support this. Either implement everything in actuator.c or everything in an extra module. As you can see i used the same logic as it is done in actuator.c so it should be interchangeable with some modifications. So far i´ve read mixed opinions regarding this in IRC.

But it would be cool to assign any PWM output / free gpio / dedicated buzzer pin to those functions (gpio buzz / pwm buzz/ pwm LED and for the future even addressable leds WS2811/WS2812 )

**Comment by tracernz** [comment]
Defer?

**Comment by mlyle** [comment]
I still have some hope of doing this.  It's a big deal to end users.  But boy are we fucked in general at the moment.

**Comment by mlyle** [comment]
@henn1001 -- thanks for your work/foundation on this .. I used your configuration for inspiration for my change.

Superseded by #1379.

**Pull request closed without merging**